doozerlib/olm/bundle.py

```diff
@@ -0,0 +1,450 @@
+import glob
+import io
+import json
+import os
+import re
+import threading
+import yaml
+
+from dockerfile_parse import DockerfileParser
+from doozerlib import brew, exectools, logutil, pushd
+
+
+class OLMBundle(object):
+    """This class is responsible for generating bundle containers out of previously built operators
+
+    Every OLM Operator image should have a corresponding bundle container, which is mostly empty,
+    carrying only the operator's manifests and some special Dockerfile labels, that allows the
+    bundle container to publish those manifests on operator's behalf
+
+    Even though each bundle container has its dedicated distgit repo, they are not meant to be
+    independently built, due to their tight coupling to corresponding operators
+    """
+
+    def __init__(self, runtime):
+        self.runtime = runtime
+
+    def rebase(self, operator_nvr):
+        """Update bundle distgit contents with manifests from given operator NVR
+        Perform image SHA replacement on manifests before commit & push
+        Annotations and Dockerfile labels are re-generated with info from operator's package YAML
+
+        :param string operator_nvr: Operator NVR (format: my-operator-v4.2.30-202004200449)
+        :return bool True if rebase succeeds, False if there was nothing new to commit
+        """
+        self.operator_nvr = operator_nvr
+        self.get_operator_buildinfo()
+        self.clone_operator()
+        self.checkout_operator_to_build_commit()
+        self.clone_bundle()
+        self.clean_bundle_contents()
+        self.get_operator_package_yaml_info()
+        self.copy_operator_manifests_to_bundle()
+        self.replace_image_references_by_sha_on_bundle_manifests()
+        self.generate_bundle_annotations()
+        self.generate_bundle_dockerfile()
+        self.create_container_yaml()
+        return self.commit_and_push_bundle(commit_msg="Update bundle manifests")
+
+    def build(self, operator_name=None):
+        """Trigger a brew build of operator's bundle
+
+        :param string operator_name: Operator name (as in ocp-build-data file name, not brew component)
+        :return bool True if build succeeds, False otherwise
+        """
+        if operator_name:
+            self.operator_repo_name = 'containers/{}'.format(operator_name)
+
+        self.clone_bundle()
+
+        if not self.trigger_bundle_container_build():
+            return False
+
+        return self.watch_bundle_container_build()
+
+    def get_latest_bundle_build(self):
+        """Get NVR of latest bundle build tagged on given target
+
+        :return string: NVR of latest bundle build
```

Comment:
```suggestion
        :return string: NVR of latest bundle build, or "" if there is none.
```
... although I'd kind of prefer `None` but not important.

Reply:
:heavy_check_mark: